### PR TITLE
MAC changes

### PR DESCRIPTION
**Pull request by markuskurbel**
An additional state responsible for the pg_hba creation would be nice as well, since the scenario where one user has multiple entries is not handled atm.

### CODE REVIEW

File: init.sls [hunk 1]
```diff
@@ -72,20 +72,31 @@ chown_pgdata:
 
 # Deploy users. Sort them, so order is not changing between salt runs
 {% for config in pillar['postgresql']['users']|default({}) %}
-createuser-{{ loop.index }}:
+createuser-{{ config['username'] }}:
```
Comment (chr4) [inline init.sls:75]: Why did you change this? `loop.index` was used here to allow multiple rules per user.

File: init.sls [hunk 2]
```diff
@@ -72,20 +72,31 @@ chown_pgdata:
...
     - name: psql -t -c "CREATE ROLE {{ config['username'] }} {{ config['options']|default('NOSUPERUSER NOCREATEDB NOCREATEROLE INHERIT LOGIN') }};"
     - unless: psql -t -c "SELECT 1 FROM pg_roles WHERE rolname='{{ config['username'] }}'" |grep -q 1
     - runas: postgres
+{% endfor %}
 
-# The "replication" keyword is not a real database, but just used for permissions in pg_hba.conf
+{% for config in pillar['postgresql']['databases']|default({}) %}
 {% if config['database'] != "replication" %}
-createdb-{{ loop.index }}:
+createdb-{{ config['database'] }}:
```
Comment (chr4) [inline init.sls:84]: I like the idea of having additional databases. Unfortunately, this would break existing configurations, were databases where created using the following pillar fields:

```yaml
users: 
  - username: test
    database: test
```

Is there a simple way to provide both syntaxes? Otherwise, a migration guide needs to be written (in README or CHANGELOG).

Another alternative would be to allow multiple databases per user, e.g. like this:

```yaml
users:
  - username: test
    databases: [bla, blubb]
```

The `database: ''` and `databases: []` pillars could even be merged automatically (not sure if this is necessary though, I suppose just supporting `databases` as an array would be sufficient).

File: init.sls [hunk 3]
```diff
@@ -72,20 +72,31 @@ chown_pgdata:
...
     - unless: psql -t -c "SELECT 1 FROM pg_database WHERE datname='{{ config['database'] }}'" |grep -q 1
     - runas: postgres
     - require:
-      - cmd: createuser-{{ loop.index }}
+      - cmd: createuser-{{ config['username'] }}
 {% endif %}
 {% endfor %}
+
+{% for config in pillar['postgresql']['schemas']|default({}) %}
+createschema-{{ config['schemaname'] }}:
```
Comment (chr4) [inline init.sls]: If I understand correctly, a schema always depends on a database. Therefore, a schema could be sub-section of the database configuration?

Something like this:

```yaml
databases:
  - database: mydb
    owner: myuser
    schema: myschema
```

You are also hard-depending on a username when creating a schema, so it might even be more consise to migrate both to the user configuration section. Maybe the user section then might be a little cluttered - we should evaluate both suggestions.

File: init.sls [hunk 4]
```diff
@@ -72,20 +72,31 @@ chown_pgdata:
...
     - runas: postgres
+{% endfor %}
 
-# The "replication" keyword is not a real database, but just used for permissions in pg_hba.conf
+{% for config in pillar['postgresql']['databases']|default({}) %}
 {% if config['database'] != "replication" %}
-createdb-{{ loop.index }}:
+createdb-{{ config['database'] }}:
   cmd.run:
     - name: psql -t -c "CREATE DATABASE {{ config['database'] }} OWNER {{ config['username'] }};"
```
Comment (chr4) [inline init.sls]: When using a seperate database configuration key, I'd uggest using the `owner` keyword instead of `username` to be more consistent with the SQL keywords.  I'm not entirely sure on this though, as we're currently using `username` instead of `role` when creating a new user as well. So this might actually be more intuitive. 
What do you think?

File: init.sls [hunk 5]
```diff
@@ -72,20 +72,31 @@ chown_pgdata:
...
     - require:
-      - cmd: createuser-{{ loop.index }}
+      - cmd: createuser-{{ config['username'] }}
 {% endif %}
 {% endfor %}
+
+{% for config in pillar['postgresql']['schemas']|default({}) %}
+createschema-{{ config['schemaname'] }}:
+  cmd.run:
+    - name: psql -t -d {{ config['database'] }} -c "CREATE SCHEMA IF NOT EXISTS {{ config['schemaname'] }} AUTHORIZATION {{ config['username'] }};"
```
Comment (chr4) [inline init.sls]: Do not use `IF NOT EXISTS`, because that would actually run the query on every salt run.
I think an `unless` clause like above would be better, because then the salt output would indicate whether a schema is actually created correctly. (This is especially useful when using `test=true`)

File: pillar.example [hunk 6]
```diff
@@ -66,3 +66,13 @@ postgresql:
       method: trust
       options: NOSUPERUSER NOCREATEDB NOCREATEROLE INHERIT LOGIN REPLICATION
       password:
+
+  databases:
+    - database: triebwerk
+      username: deploy
+      gisdb: true
```
Comment (chr4) [inline pillar.example]: I don't like the `gisdb` keyword here. I think a generic field `extensions` would be better, and would work with all sorts of extensions (especially built-ins)

```yaml
- database: triebwerk
  extensions: [postgis]
```

File: postgis.sls [hunk 7]
```diff
@@ -2,3 +2,14 @@
 postgis:
   pkg.installed:
      - pkgs: [postgresql-{{ pillar['postgresql']['version'] }}-postgis-{{ pillar['postgresql']['postgis']['version'] }}]
+
+{% for config in pillar['postgresql']['database'] %}
+  {% if config['gisdb'] %}
+postgis-extension-{{ loop.index }}:
+  cmd.run:
+    - name: psql -d {{ config['database'] }} -t -c "CREATE EXTENSION IF NOT EXISTS postgis;"
```
Comment (chr4) [inline postgis.sls]: Please also use an `unless` guard here, same reasons as pointed out when commenting `CREATE SCHEMA`.

File: reload.sls [hunk 8]
```diff
@@ -0,0 +1,9 @@
+# Reload postgres config
```
Comment (chr4) [inline reload.sls:1]: What purpose does this state have?

A database user can invoke the SQL command, and all changes should automatically be reloaded anyway when applying a state.

In case you're trying to have a state that can reload configurations on a lot of servers, I'd suggest using: `salt '*' cmd.run 'systemctl reload postgresql'` instead.

**Review by chr4 — CHANGES_REQUESTED** [review]

File: init.sls (hunk 1 above)
Comment (markuskurbel) [reply init.sls:75]: I changed this to have a well defined dependency mapping in the database creation state. I am not quite sure, what you mean by "multiple roles per user", as the role is basically the user.

Comment (chr4) [reply init.sls:75]: Sorry, I mean rules, not roles.

You need to be able to set multiple rules for a single user, like e.g. allowing two machines for replication acces:

```yaml
    - username: replicant
      type: host
      database: replication
      address: 1.2.3.4/32
      method: trust
      options: NOSUPERUSER NOCREATEDB NOCREATEROLE INHERIT LOGIN REPLICATION
    - username: replicant
      type: host
      database: replication
      address: 1.2.3.5/32
      method: trust
      options: NOSUPERUSER NOCREATEDB NOCREATEROLE INHERIT LOGIN REPLICATION
```

In your approach, this would result in two items with the identifier `username`, and therefore salt would complain.

File: init.sls (hunk 3 above)
Comment (markuskurbel) [reply init.sls]: Yes, definitely. In addition to your suggestion, a schema owner:

``` yaml
databases:
  - database: mydb
     owner: myowner
     schemas:
       - schemaname: myschema
          owner: myuser
     extensions: [foo, bar]
```

File: init.sls (hunk 4 above)
Comment (markuskurbel) [reply init.sls]: Yes, that is definitely better.

File: init.sls (hunk 5 above)
Comment (markuskurbel) [reply init.sls]: Isn't it basically the same? I mean either you get a salt log info (in the unless case) or a pg log info (in the IF NOT EXISTS case)... And either way there is SQL being executed. In case unless is False, even two queries, right?

File: pillar.example (hunk 6 above)
Comment (markuskurbel) [reply pillar.example]: Yes, like the schema definition. I will change this accordingly.

File: reload.sls (hunk 8 above)
Comment (markuskurbel) [reply reload.sls:1]: Whenever you create a user within a salt converge, the config needs to be reloaded. I just wanted to automate this. Your approach would work as well..I have no preference here.

Comment (chr4) [reply reload.sls:1]: This should already be done with the `watch` statement in line 11:

```yaml
    - watch:
      - file: /etc/postgresql/{{ version }}/main/*
```

Comment (markuskurbel) [reply reload.sls:1]: Yes, I also thought this. However in my test runs this wasn't the case.

File: init.sls (hunk 1 above)
Comment (markuskurbel) [reply init.sls:75]: I see. Yes this is lacking atm. This is what I meant with the initial comment, that the hba creaton needs some more care.

File: reload.sls (hunk 8 above)
Comment (chr4) [reply reload.sls:1]: That's very weird, I've been using this in multiple environments and never had this problem. Can you doublecheck? The reload only gets triggert if the config file actually was changed; maybe a change was applied before and then the resulting config didn't change?

File: extensions.sls [hunk 9]
```diff
@@ -0,0 +1,16 @@
+# Install postgres extensions
+postgis:
+  pkg.installed:
+     - pkgs: [postgresql-{{ pillar['postgresql']['version'] }}-postgis-{{ pillar['postgresql']['postgis']['version'] }}]
```
Comment (chr4) [inline extensions.sls:4]: Mmh, this doesn't really work - having a generic extension state, but then installing specific packages?

File: init.sls (hunk 1 above)
Comment (chr4) [reply init.sls:75]: Well, this is actually already implemented, just don't remove the `loop.index` :)

File: pillar.example [hunk 10]
```diff
@@ -45,24 +45,27 @@ postgresql:
...
     effective_cache_size: 4GB
+  
+  # These lists will be written in the pg_hba.conf
+  # Accepted shape:   
+  #  - ['host/local', 'DATABASE',  'USER',  'ADDRESS', 'METHOD']
+
+  acls:
+    - ['host', 'triebwerk', 'deploy', '10.14.38.0/24', 'trust']
+    - ['host', 'triebwerk', 'user_wwith_password', '10.14.38.0/24', 'password']
+    - ['host', 'repliction', 'replicant', '10.14.38.0/24', 'trust']
```
Comment (chr4) [inline pillar.example:56]: Mmh, why was this changed? There's no feature added and it's less readable this way and more error prone (e.g. when there's a line that doesn't have the address column)

File: pillar.example [hunk 11]
```diff
@@ -45,24 +45,27 @@ postgresql:
...
       options: NOSUPERUSER NOCREATEDB NOCREATEROLE INHERIT LOGIN REPLICATION
-      password:
+
+  databases:
+    - database: triebwerk
+      owner: deploy
+      schemas:
+      - schemaname: test
+        user: user_with_password
+      extensons: [postgis]
```
Comment (chr4) [inline pillar.example:71]: This looks pretty straight forward!
Is this actually a usecase to have a different schema user than the database owner?

What do you think of renaming the schema `user` attribute with `authorization` (matching the SQL `AUTHORIZATION` keyword)?

**Review by chr4 — CHANGES_REQUESTED** [review]

File: extensions.sls (hunk 9 above)
Comment (markuskurbel) [reply extensions.sls:4]: Yeah, kind of ugly. Howerver do we know that extensions are always:
`postgresql-{{ pillar['postgresql']['version'] }}-{{ extension }}-{{ pillar['postgresql']['postgis']['version'] }}` ?

File: pillar.example (hunk 10 above)
Comment (markuskurbel) [reply pillar.example:56]: Now you can have multiple rules for one user, as requested yesterday

File: pillar.example (hunk 11 above)
Comment (markuskurbel) [reply pillar.example:71]: Yes, then I have a better control over which user can access what resource, instead of handing over the DB owner to all our users.

Comment (chr4) [reply pillar.example:71]:
```yaml
  # NOTE: Database and user need to be existent
  schemas:
    - database: triebwerk
      schema: a
      authorization: postgres
```

**Review by chr4 — CHANGES_REQUESTED** [review]

**Comment by markuskurbel** [comment]
After a discussion, we concluded that having the schema and extension creation under salt control is not necesssary. All other requirements are covered by the current state.
=> closed
